src/test/java/com/google/googleidentity/security/LoginCheckServletTest.java

```diff
@@ -0,0 +1,129 @@
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Test {@link LoginCheckServlet}'s logic
+ */
+public class LoginCheckServletTest  {
+
+    public class UserSessionProvider implements Provider<UserSession> {
+        @Override
+        public UserSession get() {
+            UserSession userSession = new UserSession();
+            userSession.setUser(UserDetails.newBuilder()
+                    .setUsername("user")
+                    .setPassword(Hashing.sha256()
+                            .hashString("123456", Charsets.UTF_8).toString())
+                    .build());
+            return userSession;
+        }
+    }
+
+    @Test
+    public void testLoginCheckServlet_correctLoginRequest_redirectToResource()
+            throws ServletException, IOException {
+
+        UserDetailsService userDetailsService = new InMemoryUserDetailsService();
+
+        userDetailsService.addUser(UserDetails.newBuilder()
+                .setUsername("user")
+                .setPassword(Hashing.sha256()
+                        .hashString("123456", Charsets.UTF_8).toString())
+                .build());
+
+        LoginCheckServlet loginCheckServlet = new LoginCheckServlet(
+                new UserSessionProvider(), userDetailsService);
+
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        HttpServletResponse response = mock(HttpServletResponse.class);
+
+        when(request.getParameter("username")).thenReturn("user");
+        when(request.getParameter("password")).thenReturn(Hashing.sha256()
+                .hashString("123456", Charsets.UTF_8).toString());
+
+        StringWriter stringWriter = new StringWriter();
+        PrintWriter writer = new PrintWriter(stringWriter);
+        when(response.getWriter()).thenReturn(writer);
+
+        loginCheckServlet.doPost(request, response);
+
+        System.out.println(stringWriter.toString());
+
+        assertTrue(stringWriter.toString().equals("/resource/user\r\n"));
+
+    }
+
+    @Test
+    public void testLoginCheckServlet_wrongLoginRequest_redirectToLogin()
+            throws ServletException, IOException {
+
+        UserDetailsService userDetailsService = new InMemoryUserDetailsService();
+
+        userDetailsService.addUser(UserDetails.newBuilder()
+                .setUsername("user")
+                .setPassword(Hashing.sha256()
+                        .hashString("123456", Charsets.UTF_8).toString())
+                .build());
+
+        LoginCheckServlet loginCheckServlet = new LoginCheckServlet(
+                new UserSessionProvider(), userDetailsService);
+
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        HttpServletResponse response = mock(HttpServletResponse.class);
+
+        when(request.getParameter("username")).thenReturn("user");
+        when(request.getParameter("password")).thenReturn(Hashing.sha256()
+                .hashString("12345", Charsets.UTF_8).toString());
```

Comment:
Consider making the password something like "wrong password" to make it clear to the reader why this shouldn't work. Alternatively (or additionally) add a comment here to highlight that this is what causes the request to be redirected.